File: content/src/content/jcr_root/apps/core/wcm/sandbox/commons/ui/childreneditor/clientlibs/js/childreneditor.js
/*******************************************************************************
 * Copyright 2018 Adobe Systems Incorporated
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
(function($, ns, channel, window) {
    "use strict";

    var deletedItems = [];

    // Remove item
    $(document).on("coral-collection:remove", "coral-multifield", function(event) {
        var $item = $(event.detail.item);
        var path = $item.data("path");
        deletedItems.push(path);
    });

    // Add item
    $(document).on("coral-collection:add", "coral-multifield", function(event) {
        if (this === event.target) {
            // We need one more frame to make sure the item renders the template in the DOM
            Coral.commons.nextFrame(function() {
                $(this).trigger("foundation-contentloaded");
                var $item = $(event.detail.item);

                var parentPath = $item.closest("coral-multifield").data("parentPath");

                var itemName = "item_" + Date.now();
                var inputName = "./" + itemName + "/jcr:title";
                var itemPath = parentPath + "/" + itemName;
                $item.attr("data-path", itemPath);
                var $input = $item.find("input");
                $input.attr("name", inputName);
            }.bind(this));
        }
    });

    function processRequests() {
        var requests = [];

        // Process removed items
        for (var i = 0; i < deletedItems.length; i++) {
            var itemPath = deletedItems[i];
            requests.push(
                {
                    type: "POST",
                    url: itemPath,
                    data: { ":operation": "delete" }
                }
            );
        }
        deletedItems = [];

        // Process re-ordered items
        $("coral-multifield-item").each(function(idx) {
            var $item = $(this);
            var itemPath = $item.data("path");
            requests.push(
                {
                    type: "POST",
                    url: itemPath,
                    data: { ":order": idx }
                }
            );
        });

        var chain = $.when();
        requests.forEach(function(request) {
            chain = chain.then(function() {
                return $.ajax(request);
            });
        });

    }

    $(window).adaptTo("foundation-registry").register("foundation.form.submit", {
        selector: "*",
        handler: function(formEl) {

            /*
            */

            return {
                post: function() {
                    processRequests();
                }
            };
        }
    });


}(jQuery, Granite.author, jQuery(document), this));

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/sandbox/ui/ChildrenEditor.java
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2018 Adobe Systems Incorporated
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
package com.adobe.cq.wcm.core.components.sandbox.ui;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.PostConstruct;

import org.apache.commons.lang3.StringUtils;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.resource.Resource;
import org.apache.sling.api.resource.ResourceResolver;
import org.apache.sling.models.annotations.Model;
import org.apache.sling.models.annotations.injectorspecific.Self;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.day.cq.wcm.api.components.Component;
import com.day.cq.wcm.api.components.ComponentManager;

@Model(adaptables = {SlingHttpServletRequest.class})
public class ChildrenEditor {

    private static final Logger LOGGER = LoggerFactory.getLogger(ChildrenEditor.class);

    public final static String RESOURCE_TYPE = "core/wcm/sandbox/commons/ui/childreneditor";

    @Self
    private SlingHttpServletRequest request;

    private Resource container;

    private List<Resource> children;

    @PostConstruct
    private void initModel() {
        readChildren();
    }

    private void readChildren() {
        children = new ArrayList<>();
        String containerPath = request.getRequestPathInfo().getSuffix();
        if (StringUtils.isNotEmpty(containerPath)) {
            ResourceResolver resolver = request.getResourceResolver();
            container = resolver.getResource(containerPath);
            if (container != null) {
                ComponentManager componentManager = request.getResourceResolver().adaptTo(ComponentManager.class);
                for (Resource resource : container.getChildren()) {
                    if (resource != null) {
                        Component component = componentManager.getComponentOfResource(resource);
                        if (component != null) {
                            children.add(resource);
                        }
                    }
                }
            }
        }
    }

    public List<Resource> getChildren() {
        return children;
    }

    public Resource getContainer() {
        return container;
    }
}
